### DISCUSSION TITLE
Move content templates tree to content section

### DISCUSSION THREAD

**Discussion by bjarnef** [umbraco/Umbraco-CMS#11743] [Features and ideas]
It is possible to remove/filter a tree node from code and inject a custom tree node: https://our.umbraco.com/documentation/extending/section-trees/trees#treenodesrendering

but is it possible to move an existing tree node to another tree group or section? For example I think it makes sense to move content templates to content section or another section than "Settings" which we mostly don't want to grant access to editors.

Content editors are able to create content templates, but by default they can't edit or delete these.

**Comment by bjarnef** [comment]
In Umbraco 9.1.0+ it is possible to remove the existing "Content Templates" tree:

```
public class TreeComposer : IComposer
{
    public void Compose(IUmbracoBuilder builder)
    {
        builder.Trees().RemoveTreeController<ContentBlueprintTreeController>();
    }
}
```

and then register a new controller inheriting from `ContentBlueprintTreeController` which is auto-registered when decorated with `Tree` attribute.

```
[Authorize(Policy = AuthorizationPolicies.SectionAccessContent)]
[Tree(Constants.Applications.Content, Constants.Trees.ContentBlueprints)] //TreeGroup = Constants.Trees.Groups.Settings
[PluginController(Constants.Web.Mvc.BackOfficeTreeArea)]
[CoreTree]

public class ContentTemplatesController : ContentBlueprintTreeController
{
    public ContentTemplatesController(
        ILocalizedTextService localizedTextService,
        UmbracoApiControllerTypeCollection umbracoApiControllerTypeCollection,
        IMenuItemCollectionFactory menuItemCollectionFactory,
        IContentService contentService,
        IContentTypeService contentTypeService,
        IEntityService entityService,
        IEventAggregator eventAggregator) 
            : base(localizedTextService, umbracoApiControllerTypeCollection, menuItemCollectionFactory, 
                  contentService, contentTypeService, entityService, eventAggregator)
    {
        
    }

    protected override ActionResult<TreeNode> CreateRootNode(FormCollection queryStrings)
    {
        return base.CreateRootNode(queryStrings);
    }
}
```

This will create two root tree nodes "Content" and "Content Templates".

![image](https://user-images.githubusercontent.com/2919859/147488707-3616c2d1-af87-47cf-944b-f55d59d91b47.png)

If using `TreeGroup = Constants.Trees.Groups.Settings` "Content Templates" tree is moved to a "Settings" tree group while existing "Content" tree is moved to a "Third Party" tree group, which is a bit confusing. Maybe tree controllers decorated with `CoreTree` could be default be added to a "Content" tree group instead or a default tree group in section, e.g. "Content", "Media" etc.?

![image](https://user-images.githubusercontent.com/2919859/147489036-85d77aa6-c302-41fd-82ae-923f166de16b.png)

**Comment by bjarnef** [comment]
One isssue I noticed when moving the Content Templates tree to content section is that the `GetAll()` method fails.

![image](https://user-images.githubusercontent.com/2919859/147501329-92854809-1261-4b4f-bd4b-64c285e92bd9.png)

but when granting "Editors" group access to "Settings" section as well this doesn't fail.
I guess it fails fetching content types unless the user has access to settings (or the document types tree node).

https://github.com/umbraco/Umbraco-CMS/blob/v9/contrib/src/Umbraco.Web.BackOffice/Controllers/ContentTypeController.cs#L436

**Comment by bjarnef** [comment]
Furthermore I noticed the "Create Content Template" menu action doesn't do anything unless the user has access to settings section or this basically route to `/umbraco#/settings/contentBlueprints/edit/-1?create=true&doctype=standardPage` although the tree is moved to content section, but it seems it is possible to override this as well to  be `/umbraco#/content/contentBlueprints/edit/-1?create=true&doctype=standardPage` instead.

```
protected override ActionResult<MenuItemCollection> GetMenuForNode(string id, FormCollection queryStrings)
{
    var collection = base.GetMenuForNode(id, queryStrings);
            
    if (collection.Value.Items.Count > 0)
    {
        foreach (var menuItem in collection.Value.Items)
        {
            if (menuItem.Action is ActionCreateBlueprintFromContent)
            {
                var route = menuItem.AdditionalData["actionRoute"]?.ToString();

                if (!string.IsNullOrEmpty(route))
                {
                    menuItem.NavigateToRoute(route.Replace("/settings/", "/content/"));
                }
            }
        }
    }

    return collection;
}
```


## Final solution for now

```
public class TreeComposer : IComposer
{
    public void Compose(IUmbracoBuilder builder)
    {
        builder.Trees().RemoveTreeController<ContentBlueprintTreeController>();
    }
}
```

```
[Authorize(Policy = AuthorizationPolicies.SectionAccessContent)]
[Tree(Constants.Applications.Content, Constants.Trees.ContentBlueprints, SortOrder = 10)] //TreeGroup = Constants.Trees.Groups.Settings
[PluginController(Constants.Web.Mvc.BackOfficeTreeArea)]
[CoreTree]
public class ContentTemplatesController : ContentBlueprintTreeController
{
    public ContentTemplatesController(
        ILocalizedTextService localizedTextService,
        UmbracoApiControllerTypeCollection umbracoApiControllerTypeCollection,
        IMenuItemCollectionFactory menuItemCollectionFactory,
        IContentService contentService,
        IContentTypeService contentTypeService,
        IEntityService entityService,
        IEventAggregator eventAggregator)
        : base(localizedTextService, umbracoApiControllerTypeCollection, menuItemCollectionFactory,
            contentService, contentTypeService, entityService, eventAggregator)
    {
    }

    protected override ActionResult<TreeNode> CreateRootNode(FormCollection queryStrings)
    {
        var root = base.CreateRootNode(queryStrings);

        if (!string.IsNullOrEmpty(root.Value.RoutePath))
        {
            // Update route to content section
            root.Value.RoutePath = root.Value.RoutePath.Replace(Constants.Applications.Settings, Constants.Applications.Content);
        }

        return root;
    }

    protected override ActionResult<MenuItemCollection> GetMenuForNode(string id, FormCollection queryStrings)
    {
        var collection = base.GetMenuForNode(id, queryStrings);

        if (collection.Value.Items.Count > 0)
        {
            foreach (var menuItem in collection.Value.Items)
            {
                if (menuItem.Action is ActionCreateBlueprintFromContent)
                {
                    var route = menuItem.AdditionalData["actionRoute"]?.ToString();

                    if (!string.IsNullOrEmpty(route))
                    {
                        // Update route to content section
                        menuItem.NavigateToRoute(route.Replace($"/{Constants.Applications.Settings}/", $"/{Constants.Applications.Content}/"));
                    }
                }
            }
        }

        return collection;
    }
}
```

**Comment by bjarnef** [comment]
@AndyButland not sure if this is planned for Umbraco 18, but it may be worth considering moving content templates to library section by default.

If it should still be present in Settings section, then probably allow the tree to be removed from a section and inserted into another section.
I haven't checked it this has changed in the new backoffice, but previously the workaround was to remove core tree and register a custom tree (inheriting from `ContentBlueprintTreeController`).

**Reply by skttl** [reply]
IIRC it was mentioned in the winter keynote, that Content Templates would be part of the new Library section.

**Reply by bjarnef** [reply]
I recall it has been mentioned, just wanted to check as this discussion then can be closed :)
